### PR TITLE
nixos/docker: ensure ipv4 forwarding is enabled

### LINKED ISSUE #118656: virtualisation/docker: net.ipv4.conf.default.forwarding not enabled in sysctl.d
**Describe the bug**
Upon enabling docker via `virtualisation.docker.enable = true` sysctl.d is not updated to allow IP forwarding which is required for containers to communicate with outside world.

**To Reproduce**
1. configure `virtualisation.docker.enable = true`
2.  ``` # nixos-rebuild switch```
This will get `dockerd` up and running with `net.ipv4.conf.default.forwarding` set to `1` which in my understanding is set by the `dockerd` itself.
3. reload `systemd-sysctl` to apply system-wide configured sysctl values and check `net.ipv4.conf.all.forwarding`
```
# systemctl restart systemd-sysctl
# sysctl net.ipv4.conf.default.forwarding
net.ipv4.conf.default.forwarding = 0
```
By now your docker containers lost connectivity.

**Expected behavior**
I expect `virtualisation/docker.nix` to override relevant `boot.kernel.sysctl` values. Something like this:
```
boot = {
  kernel.sysctl = {
    "net.ipv4.conf.all.forwarding" = mkOverride 99 true;
    "net.ipv4.conf.default.forwarding" = mkOverride 99 true;
  }
}
```

**Additional context**
I discovered this problem when applying other sysctl values on a running system without a reboot.

**Notify maintainers**
@ryneeverett @bachp @rasendubi 

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.108, NixOS, 20.09.3732.91b77fe6942 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-20.09.3732.91b77fe6942"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by alekna**
###### Motivation for this change
Fixes #118656

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/virtualisation/docker.nix [hunk 1]
```diff
@@ -150,6 +150,10 @@ in
 
   config = mkIf cfg.enable (mkMerge [{
       boot.kernelModules = [ "bridge" "veth" ];
+      boot.kernel.sysctl = {
+        "net.ipv4.conf.all.forwarding" = mkOverride 99 true;
+        "net.ipv4.conf.default.forwarding" = mkOverride 99 true;
```
Comment (mohe2015) [inline nixos/modules/virtualisation/docker.nix:155]: https://docs.docker.com/network/bridge/#enable-forwarding-from-docker-containers-to-the-outside-world was the only upstream source I could find. I also checked by manually restarting docker after doing a system rebuild and comparing sysctl before and after and these two options are definitely set by docker on the host.

https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt specifies what they mean if anybody wants to read that.

**Review by mohe2015 — APPROVED** [review]
I manually confirmed that these options are set by docker and are also not persisted properly when reloading `systemd-sysctl`.

Also @alekna you provided great steps for easily reproducing the bug.

I also manually confirmed that this fixes the problem. I would consider this a proper fix although I must say that I don't know docker nor networking that well.

Edit: I think this should get into the next stable release as this would break docker subtly on rebuilding your system.

**Review by lukegb — CHANGES_REQUESTED** [review]
Thanks for the contribution! The commit message should be `nixos/docker` and not `docker`, since this is a fix to the module rather than the package.

**Review by lukegb — APPROVED** [review]

**Review by mohe2015 — APPROVED** [review]

**Comment by misuzu** [comment]
I believe this change broke evaluation when both docker and NAT are enabled, so this:

```nix
{
  networking.nat.enable = true;
  virtualisation.docker.enable = true;
}
```
leads to this:
```
error: The unique option `boot.kernel.sysctl.net.ipv4.conf.all.forwarding' is defined multiple times. Definition values:
       - In `/nix/store/fh7qqd2j7zkhfjzj31njl9k66b0fl3p3-source/nixos/modules/virtualisation/docker.nix': true
       - In `/nix/store/fh7qqd2j7zkhfjzj31njl9k66b0fl3p3-source/nixos/modules/services/networking/nat.nix': true
(use '--show-trace' to show detailed location information)
```

**Comment by mohe2015** [comment]
I think this should be reverted for now then. I am thinking if the following change would be useful though (I actually have little knowledge of the internals of the options system):
```patch
diff --git a/nixos/modules/config/sysctl.nix b/nixos/modules/config/sysctl.nix
index e59c7a32c28..a5ee435c956 100644
--- a/nixos/modules/config/sysctl.nix
+++ b/nixos/modules/config/sysctl.nix
@@ -11,7 +11,7 @@ let
         checkType = x: isBool x || isString x || isInt x || x == null;
       in
         checkType val || (val._type or "" == "override" && checkType val.content);
-    merge = loc: defs: mergeOneOption loc (filterOverrides defs);
+    merge = loc: defs: mergeEqualOption loc (filterOverrides defs);
   };
 
 in
```

**Comment by alekna** [comment]
if it is easy to fix, why reverting?

**Comment by mohe2015** [comment]
> if it is easy to fix, why reverting?

Because I don't know of the side effects of my proposed fix but feel free to create the PR with my proposed change and maybe somebody knows whether that's a valid fix.

**Comment by alekna** [comment]
Thanks @mohe2015, I will do some research into it and then submit a PR. I kind of feel responsible for fixing this issue. I will then tag you in for review, if you don't mind.

**Comment by mohe2015** [comment]
Sure. https://github.com/NixOS/nixpkgs/pull/122810 is also a proposed fix that doesn't seems to have been linked.

**Pull request merged**